src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -0,0 +1,302 @@
+        /// </summary>
+        public Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>> GetDirectDependencies(CallGraphNode callerSpec)
+        {
+            if (_Dependencies.TryGetValue(callerSpec, out var deps))
+            {
+                return deps;
+            }
+            else
+            {
+                return new Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>>();
+            }
+        }
+
+        /// <summary>
+        /// Returns all specializations that are used directly within the given caller, whether they are
+        /// called, partially applied, or assigned. Each key in the returned dictionary represents a
+        /// specialization that is used by the caller. Each value in the dictionary is an array of edges
+        /// representing all the different ways the given caller specialization took a dependency on the
+        /// specialization represented by the associated key.
+        /// </summary>
+        public Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>> GetDirectDependencies(QsSpecialization callerSpec) =>
+            GetDirectDependencies(new CallGraphNode { CallableName = callerSpec.Parent, Kind = callerSpec.Kind, TypeArgs = RemovePositionFromTypeArgs(callerSpec.TypeArguments) });
+
+        // ToDo: this method needs a way of resolving type parameters before it can be completed
+        public Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>> GetAllDependencies(CallGraphNode callerSpec)
+        {
+            return new Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>>();
+
+            //HashSet<(CallGraphNode, CallGraphEdge)> WalkDependencyTree(CallGraphNode root, HashSet<(CallGraphNode, CallGraphEdge)> accum, DependencyType parentDepType)
+            //{
+            //    if (_Dependencies.TryGetValue(root, out var next))
+            //    {
+            //        foreach (var k in next)
+            //        {
+            //            // Get the maximum type of dependency between the parent dependency type and the current dependency type
+            //            var maxDepType = k.Item2.CompareTo(parentDepType) > 0 ? k.Item2 : parentDepType;
+            //            if (accum.Add((k.Item1, maxDepType)))
+            //            {
+            //                // ToDo: this won't work once Type specialization are implemented
+            //                var noTypeParams = new CallGraphNode { CallableName = k.Item1.CallableName, Kind = k.Item1.Kind, TypeArgs = QsNullable<ImmutableArray<ResolvedType>>.Null };
+            //                WalkDependencyTree(noTypeParams, accum, maxDepType);
+            //            }
+            //        }
+            //    }
+            //
+            //    return accum;
+            //}
+            //
+            //return WalkDependencyTree(callerSpec, new HashSet<(CallGraphNode, DependencyType)>(), DependencyType.NoTypeParameters).ToImmutableArray();
```

Comment:
Does all of the code need to be commented out, or just the part that isn't implemented yet (resolving type parameters)?

Reply:
I'd rather keep this around for now as it might help with implementing the method in the future, once we have the proper tools for doing this. It won't go into the master branch like this.